File: src/engine/declension.rs
use fancy_regex::Regex;
use std::collections::HashMap;

pub fn apply_natva(word: &str) -> String {
    let pattern = Regex::new(r"([rfFz][aAiIuUfFxXeEoOkKgGNpPbBmyvhM]*)n(?=[aAiIuUfFxXeEoOmyv])").unwrap();
    let mut current = word.to_string();
    while pattern.is_match(&current).unwrap_or(false) {
        current = pattern.replace(&current, "${1}R").into_owned();
    }
    current
}

fn build_grid(pr: [&str;3], dv: [&str;3], tr: [&str;3], ca: [&str;3], pa: [&str;3], sa: [&str;3], sap: [&str;3], sam: [&str;3]) -> HashMap<String, Vec<String>> {
    let mut map = HashMap::new();
    map.insert("prathama".to_string(), vec![pr[0].to_string(), pr[1].to_string(), pr[2].to_string()]);
    map.insert("dvitiya".to_string(), vec![dv[0].to_string(), dv[1].to_string(), dv[2].to_string()]);
    map.insert("tritiya".to_string(), vec![tr[0].to_string(), tr[1].to_string(), tr[2].to_string()]);
    map.insert("caturthi".to_string(), vec![ca[0].to_string(), ca[1].to_string(), ca[2].to_string()]);
    map.insert("panchami".to_string(), vec![pa[0].to_string(), pa[1].to_string(), pa[2].to_string()]);
    map.insert("sasthi".to_string(), vec![sa[0].to_string(), sa[1].to_string(), sa[2].to_string()]);
    map.insert("saptami".to_string(), vec![sap[0].to_string(), sap[1].to_string(), sap[2].to_string()]);
    map.insert("sambodhana".to_string(), vec![sam[0].to_string(), sam[1].to_string(), sam[2].to_string()]);
    map
}

pub fn decline_noun(base: &str, gender: &str) -> Result<HashMap<String, Vec<String>>, String> {
    if let Some(irreg) = crate::engine::irregulars::decline_irregular(base, gender) {
        let mut fixed = HashMap::new();
        for (k, v) in irreg {
            let natva_v: Vec<String> = v.into_iter().map(|w| apply_natva(&w)).collect();
            fixed.insert(k, natva_v);
        }
        return Ok(fixed);
    }

    let mut res = HashMap::new();
    let s = if base.len() > 0 { &base[..base.len()-1] } else { "" };
    
    if base.ends_with('a') {
        if gender == "masculine" { res = build_grid([&format!("{}aH",s), &format!("{}O",s), &format!("{}AH",s)], [&format!("{}am",s), &format!("{}O",s), &format!("{}An",s)], [&format!("{}ena",s), &format!("{}AByAm",s), &format!("{}EH",s)], [&format!("{}Aya",s), &format!("{}AByAm",s), &format!("{}eByaH",s)], [&format!("{}At",s), &format!("{}AByAm",s), &format!("{}eByaH",s)], [&format!("{}asya",s), &format!("{}ayoH",s), &format!("{}AnAm",s)], [&format!("{}e",s), &format!("{}ayoH",s), &format!("{}ezu",s)], [&format!("{}a",s), &format!("{}O",s), &format!("{}AH",s)]); }
        else if gender == "neuter" { res = build_grid([&format!("{}am",s), &format!("{}e",s), &format!("{}Ani",s)], [&format!("{}am",s), &format!("{}e",s), &format!("{}Ani",s)], [&format!("{}ena",s), &format!("{}AByAm",s), &format!("{}EH",s)], [&format!("{}Aya",s), &format!("{}AByAm",s), &format!("{}eByaH",s)], [&format!("{}At",s), &format!("{}AByAm",s), &format!("{}eByaH",s)], [&format!("{}asya",s), &format!("{}ayoH",s), &format!("{}AnAm",s)], [&format!("{}e",s), &format!("{}ayoH",s), &format!("{}ezu",s)], [&format!("{}a",s), &format!("{}e",s), &format!("{}Ani",s)]); }
    } else if base.ends_with('A') {
        if gender == "feminine" { res = build_grid([&format!("{}A",s), &format!("{}e",s), &format!("{}AH",s)], [&format!("{}Am",s), &format!("{}e",s), &format!("{}AH",s)], [&format!("{}ayA",s), &format!("{}AByAm",s), &format!("{}ABiH",s)], [&format!("{}AyE",s), &format!("{}AByAm",s), &format!("{}AByaH",s)], [&format!("{}AyAH",s), &format!("{}AByAm",s), &format!("{}AByaH",s)], [&format!("{}AyAH",s), &format!("{}ayoH",s), &format!("{}AnAm",s)], [&format!("{}AyAm",s), &format!("{}ayoH",s), &format!("{}Asu",s)], [&format!("{}e",s), &format!("{}e",s), &format!("{}AH",s)]); }
    } else if base.ends_with('I') {
        if gender == "feminine" { res = build_grid([&format!("{}I",s), &format!("{}yO",s), &format!("{}yaH",s)], [&format!("{}Im",s), &format!("{}yO",s), &format!("{}IH",s)], [&format!("{}yA",s), &format!("{}IByAm",s), &format!("{}IBiH",s)], [&format!("{}yE",s), &format!("{}IByAm",s), &format!("{}IByaH",s)], [&format!("{}yAH",s), &format!("{}IByAm",s), &format!("{}IByaH",s)], [&format!("{}yAH",s), &format!("{}yoH",s), &format!("{}InAm",s)], [&format!("{}yAm",s), &format!("{}yoH",s), &format!("{}Izu",s)], [&format!("{}i",s), &format!("{}yO",s), &format!("{}yaH",s)]); }
    } else if base.ends_with('i') {
        if gender == "masculine" { res = build_grid([&format!("{}iH",s), &format!("{}I",s), &format!("{}ayaH",s)], [&format!("{}im",s), &format!("{}I",s), &format!("{}In",s)], [&format!("{}inA",s), &format!("{}iByAm",s), &format!("{}iBiH",s)], [&format!("{}aye",s), &format!("{}iByAm",s), &format!("{}iByaH",s)], [&format!("{}eH",s), &format!("{}iByAm",s), &format!("{}iByaH",s)], [&format!("{}eH",s), &format!("{}yoH",s), &format!("{}InAm",s)], [&format!("{}O",s), &format!("{}yoH",s), &format!("{}izu",s)], [&format!("{}e",s), &format!("{}I",s), &format!("{}ayaH",s)]); }
        else if gender == "feminine" { res = build_grid([&format!("{}iH",s), &format!("{}I",s), &format!("{}ayaH",s)], [&format!("{}im",s), &format!("{}I",s), &format!("{}IH",s)], [&format!("{}yA",s), &format!("{}iByAm",s), &format!("{}iBiH",s)], [&format!("{}yE",s), &format!("{}iByAm",s), &format!("{}iByaH",s)], [&format!("{}yAH",s), &format!("{}iByAm",s), &format!("{}iByaH",s)], [&format!("{}yAH",s), &format!("{}yoH",s), &format!("{}InAm",s)], [&format!("{}yAm",s), &format!("{}yoH",s), &format!("{}izu",s)], [&format!("{}e",s), &format!("{}I",s), &format!("{}ayaH",s)]); }
    } else if base.ends_with("at") {
        let s2 = &base[..base.len()-2];
        if gender == "masculine" { res = build_grid([&format!("{}an",s2), &format!("{}antO",s2), &format!("{}antaH",s2)], [&format!("{}antam",s2), &format!("{}antO",s2), &format!("{}ataH",s2)], [&format!("{}atA",s2), &format!("{}adByAm",s2), &format!("{}adBiH",s2)], [&format!("{}ate",s2), &format!("{}adByAm",s2), &format!("{}adByaH",s2)], [&format!("{}ataH",s2), &format!("{}adByAm",s2), &format!("{}adByaH",s2)], [&format!("{}ataH",s2), &format!("{}atoH",s2), &format!("{}atAm",s2)], [&format!("{}ati",s2), &format!("{}atoH",s2), &format!("{}atsu",s2)], [&format!("{}an",s2), &format!("{}antO",s2), &format!("{}antaH",s2)]); }
    } else if base.ends_with("an") {
        let s2 = &base[..base.len()-2];
        if gender == "masculine" { res = build_grid([&format!("{}A",s2), &format!("{}AnO",s2), &format!("{}AnaH",s2)], [&format!("{}Anam",s2), &format!("{}AnO",s2), &format!("{}naH",s2)], [&format!("{}nA",s2), &format!("{}aByAm",s2), &format!("{}aBiH",s2)], [&format!("{}ne",s2), &format!("{}aByAm",s2), &format!("{}aByaH",s2)], [&format!("{}naH",s2), &format!("{}aByAm",s2), &format!("{}aByaH",s2)], [&format!("{}naH",s2), &format!("{}noH",s2), &format!("{}nAm",s2)], [&format!("{}ni",s2), &format!("{}noH",s2), &format!("{}asu",s2)], [&format!("{}an",s2), &format!("{}AnO",s2), &format!("{}AnaH",s2)]); }
    }
    
    if res.is_empty() {
        return Err(format!("Declension logic for '{}' in gender '{}' is not fully implemented in Rust yet.", base, gender));
    }
    
    for forms in res.values_mut() {
        for word in forms.iter_mut() {
            *word = apply_natva(word);
        }
    }
    
    Ok(res)
}

File: engine/declension.go
package engine

import (
	"fmt"
	"regexp"
	"strings"
)

// natvaRegex converts 'n' -> 'R' after triggering consonants (r, f, F, z)
// Note: Go's regexp package does not support lookaheads (?=...), so we capture
// the following vowel/consonant in Group 2 and replace with ${1}R${2}.
var natvaRegex = regexp.MustCompile(`([rfFz][aAiIuUfFxXeEoOkKgGNpPbBmyvhM]*)n([aAiIuUfFxXeEoOmyv])`)

func applyNatva(word string) string {
	current := word
	for natvaRegex.MatchString(current) {
		current = natvaRegex.ReplaceAllString(current, "${1}R${2}")
	}
	return current
}

func buildGrid(pr, dv, tr, ca, pa, sa, sap, sam [3]string) map[string][]string {
	return map[string][]string{
		"prathama":   {pr[0], pr[1], pr[2]},
		"dvitiya":    {dv[0], dv[1], dv[2]},
		"tritiya":    {tr[0], tr[1], tr[2]},
		"caturthi":   {ca[0], ca[1], ca[2]},
		"panchami":   {pa[0], pa[1], pa[2]},
		"sasthi":     {sa[0], sa[1], sa[2]},
		"saptami":    {sap[0], sap[1], sap[2]},
		"sambodhana": {sam[0], sam[1], sam[2]},
	}
}

// DeclineNoun generates the 8x3 declension grid for a nominal stem
func DeclineNoun(base string, gender string) (map[string][]string, error) {
	if irreg, ok := declineIrregular(base, gender); ok {
		fixed := make(map[string][]string)
		for k, v := range irreg {
			natvaV := make([]string, len(v))
			for i, w := range v {
				natvaV[i] = applyNatva(w)
			}
			fixed[k] = natvaV
		}
		return fixed, nil
	}

	res := make(map[string][]string)
	s := ""
	if len(base) > 0 {
		s = base[:len(base)-1]
	}

	if strings.HasSuffix(base, "a") {
		if gender == "masculine" {
			res = buildGrid(
				[3]string{s + "aH", s + "O", s + "AH"},
				[3]string{s + "am", s + "O", s + "An"},
				[3]string{s + "ena", s + "AByAm", s + "EH"},
				[3]string{s + "Aya", s + "AByAm", s + "eByaH"},
				[3]string{s + "At", s + "AByAm", s + "eByaH"},
				[3]string{s + "asya", s + "ayoH", s + "AnAm"},
				[3]string{s + "e", s + "ayoH", s + "ezu"},
				[3]string{s + "a", s + "O", s + "AH"},
			)
		} else if gender == "neuter" {
			res = buildGrid(
				[3]string{s + "am", s + "e", s + "Ani"},
				[3]string{s + "am", s + "e", s + "Ani"},
				[3]string{s + "ena", s + "AByAm", s + "EH"},
				[3]string{s + "Aya", s + "AByAm", s + "eByaH"},
				[3]string{s + "At", s + "AByAm", s + "eByaH"},
				[3]string{s + "asya", s + "ayoH", s + "AnAm"},
				[3]string{s + "e", s + "ayoH", s + "ezu"},
				[3]string{s + "a", s + "e", s + "Ani"},
			)
		}
	} else if strings.HasSuffix(base, "A") {
		if gender == "feminine" {
			res = buildGrid(
				[3]string{s + "A", s + "e", s + "AH"},
				[3]string{s + "Am", s + "e", s + "AH"},
				[3]string{s + "ayA", s + "AByAm", s + "ABiH"},
				[3]string{s + "AyE", s + "AByAm", s + "AByaH"},
				[3]string{s + "AyAH", s + "AByAm", s + "AByaH"},
				[3]string{s + "AyAH", s + "ayoH", s + "AnAm"},
				[3]string{s + "AyAm", s + "ayoH", s + "Asu"},
				[3]string{s + "e", s + "e", s + "AH"},
			)
		}
	} else if strings.HasSuffix(base, "I") {
		if gender == "feminine" {
			res = buildGrid(
				[3]string{s + "I", s + "yO", s + "yaH"},
				[3]string{s + "Im", s + "yO", s + "IH"},
				[3]string{s + "yA", s + "IByAm", s + "IBiH"},
				[3]string{s + "yE", s + "IByAm", s + "IByaH"},
				[3]string{s + "yAH", s + "IByAm", s + "IByaH"},
				[3]string{s + "yAH", s + "yoH", s + "InAm"},
				[3]string{s + "yAm", s + "yoH", s + "Izu"},
				[3]string{s + "i", s + "yO", s + "yaH"},
			)
		}
	} else if strings.HasSuffix(base, "i") {
		if gender == "masculine" {
			res = buildGrid(
				[3]string{s + "iH", s + "I", s + "ayaH"},
				[3]string{s + "im", s + "I", s + "In"},
				[3]string{s + "inA", s + "iByAm", s + "iBiH"},
				[3]string{s + "aye", s + "iByAm", s + "iByaH"},
				[3]string{s + "eH", s + "iByAm", s + "iByaH"},
				[3]string{s + "eH", s + "yoH", s + "InAm"},
				[3]string{s + "O", s + "yoH", s + "izu"},
				[3]string{s + "e", s + "I", s + "ayaH"},
			)
		} else if gender == "feminine" {
			res = buildGrid(
				[3]string{s + "iH", s + "I", s + "ayaH"},
				[3]string{s + "im", s + "I", s + "IH"},
				[3]string{s + "yA", s + "iByAm", s + "iBiH"},
				[3]string{s + "yE", s + "iByAm", s + "iByaH"},
				[3]string{s + "yAH", s + "iByAm", s + "iByaH"},
				[3]string{s + "yAH", s + "yoH", s + "InAm"},
				[3]string{s + "yAm", s + "yoH", s + "izu"},
				[3]string{s + "e", s + "I", s + "ayaH"},
			)
		}
	} else if strings.HasSuffix(base, "at") {
		s2 := ""
		if len(base) >= 2 {
			s2 = base[:len(base)-2]
		}
		if gender == "masculine" {
			res = buildGrid(
				[3]string{s2 + "an", s2 + "antO", s2 + "antaH"},
				[3]string{s2 + "antam", s2 + "antO", s2 + "ataH"},
				[3]string{s2 + "atA", s2 + "adByAm", s2 + "adBiH"},
				[3]string{s2 + "ate", s2 + "adByAm", s2 + "adByaH"},
				[3]string{s2 + "ataH", s2 + "adByAm", s2 + "adByaH"},
				[3]string{s2 + "ataH", s2 + "atoH", s2 + "atAm"},
				[3]string{s2 + "ati", s2 + "atoH", s2 + "atsu"},
				[3]string{s2 + "an", s2 + "antO", s2 + "antaH"},
			)
		}
	} else if strings.HasSuffix(base, "an") {
		s2 := ""
		if len(base) >= 2 {
			s2 = base[:len(base)-2]
		}
		if gender == "masculine" {
			res = buildGrid(
				[3]string{s2 + "A", s2 + "AnO", s2 + "AnaH"},
				[3]string{s2 + "Anam", s2 + "AnO", s2 + "naH"},
				[3]string{s2 + "nA", s2 + "aByAm", s2 + "aBiH"},
				[3]string{s2 + "ne", s2 + "aByAm", s2 + "aByaH"},
				[3]string{s2 + "naH", s2 + "aByAm", s2 + "aByaH"},
				[3]string{s2 + "naH", s2 + "noH", s2 + "nAm"},
				[3]string{s2 + "ni", s2 + "noH", s2 + "asu"},
				[3]string{s2 + "an", s2 + "AnO", s2 + "AnaH"},
			)
		}
	}

	if len(res) == 0 {
		return nil, fmt.Errorf("declension logic for '%s' in gender '%s' is not implemented", base, gender)
	}

	for k, forms := range res {
		for i, word := range forms {
			forms[i] = applyNatva(word)
		}
		res[k] = forms
	}

	return res, nil
}